lib/packs/private.rb

```diff
@@ -477,45 +477,91 @@ def self.packwerk_package_to_pack(package)
       Packs.find(package.name)
     end
 
-    sig { params(packs: T::Array[Packs::Pack]).void }
-    def self.get_info(packs: Packs.all)
-      inbound_violations = {}
-      outbound_violations = {}
+    sig do
+      params(
+        packs: T::Array[Packs::Pack],
+        format: Symbol,
+        types: T::Array[Symbol],
+        include_date: T::Boolean
+      ).void
+    end
+    def self.get_info(packs: Packs.all, format: :detail, types: %i[privacy dependency architecture], include_date: false)
+      require 'csv' if format == :csv
+
+      today = Date.today.iso8601
+      violations = {
+        inbound: {},
+        outbound: {}
+      }
+      directions = violations.keys
+      dir_x_types = directions.product(types)
+
       ParsePackwerk.all.each do |p|
         p.violations.each do |violation|
-          outbound_violations[p.name] ||= []
-          outbound_violations[p.name] << violation
-          inbound_violations[violation.to_package_name] ||= []
-          inbound_violations[violation.to_package_name] << violation
+          violations[:outbound][p.name] ||= []
+          violations[:outbound][p.name] << violation
+          violations[:inbound][violation.to_package_name] ||= []
+          violations[:inbound][violation.to_package_name] << violation
         end
       end
 
-      all_inbound = T.let([], T::Array[ParsePackwerk::Violation])
-      all_outbound = T.let([], T::Array[ParsePackwerk::Violation])
+      all = {
+        inbound: T.let([], T::Array[ParsePackwerk::Violation]),
+        outbound: T.let([], T::Array[ParsePackwerk::Violation])
+      }
       packs.each do |pack|
-        all_inbound += inbound_violations[pack.name] || []
-        all_outbound += outbound_violations[pack.name] || []
+        all[:inbound] += violations[:inbound][pack.name] || []
+        all[:outbound] += violations[:outbound][pack.name] || []
       end
 
-      puts "There are #{all_inbound.select(&:privacy?).sum { |v| v.files.count }} total inbound privacy violations"
-      puts "There are #{all_inbound.select(&:dependency?).sum { |v| v.files.count }} total inbound dependency violations"
-      puts "There are #{all_outbound.select(&:privacy?).sum { |v| v.files.count }} total outbound privacy violations"
-      puts "There are #{all_outbound.select(&:dependency?).sum { |v| v.files.count }} total outbound dependency violations"
```

Comment:
I find the original code with the 4 puts easier to read and understand. I understand that there's duplication. I would prefer a small amount of duplication over an abstraction layer that makes it harder for new-to-the-code-base developer to understand. 

If you agree, that would mean that we don't need these above:
```
      violations = {
        inbound: {},
        outbound: {}
      }
      directions = violations.keys
      dir_x_types = directions.product(types)
```
And we could simply the printing below.

Otherwise, I'm happy to merge this PR.

Reply:
Thanks for the review!

The problem with that is that then we can't easily support new violation types. Or even allow the user to select which ones they want from the CLI.

In my case, I'm currently interested in `architecture` violations and not `dependency` nor `privacy` for the time being.

An intermediate solution would be to have duplicate lines one for inbound and one for outbound, but still loop through a list of violation types. Would you be ok with that?

Reply:
How about now, @professor? I removed the `directions` and `dir_x_types` variables and the looping over `directions`. I can't really remove the looping over `types` as that's one of the main purposes of this PR.

Reply:
Thanks for the update to the PR. 

When I look at the original code, I would find something like [1] easier to read than [2]


```
# [1]
puts "There are #{all_inbound.select(&:privacy?).sum { |v| v.files.count }} total inbound privacy violations" if types.include?(:privacy)
puts "There are #{all_inbound.select(&:dependency?).sum { |v| v.files.count }} total inbound dependency violations" if types.include?(:dependency)
puts "There are #{all_inbound.select(&: architecture?).sum { |v| v.files.count }} total inbound dependency violations" if types.include?(: architecture)
```

```
# [2]
types.each do |type|
  key = ['inbound', type, 'violations'].join('_').to_sym
  row[key] = (violations[:inbound][pack.name] || []).select { _1.type.to_sym == type }.sum { |v| v.files.count }
end
...
types.each do |type|
    key = ['inbound', type, 'violations'].join('_').to_sym
    puts "There are #{row[key]} inbound #{type} violations"
end
```

[1] will allow the user to select which types they want displayed. Maybe I'm missing something else with the change. Thanks again @oboxodo for improving the code with this feature.

Reply:
Thanks for the feedback, @professor. Here are my thoughts:

Yes, [1] is more readable than [2]. But [1] is hardcoded to only support `privacy`, `dependency`, and `architecture`. IIRC I've seen `visibility` checks too in packwerk and then that code would need to be changed in order to support that (or any other future check FWIW). [2] has the benefit of not needing to be updated because it supports _any_ checks that could come in the future (as long as they support the same API as the current ones which seems to be the goal).

On top of that, you're showing [2] as two separate loops but those loops are there for different reasons. The goal of the first loop is to collect the violations _regardless_ of how they're going to be reported. I introduced that `row` hash as an intermediate format-independent way to collect the data, then there's separate code to interpret that row and convert it into either the `csv` format or the `detail` (original) format.

I've actually thought a follow-up PR could make this separation more explicit and then the current method would be left only with the pure data-collection code and then delegate outputting to different methods depending on the selected format.

I'd really like to avoid hardcoding anything related to the violation types. Thanks to this, [this](https://github.com/rubyatscale/packs/pull/110/commits/8c874fe58f3e7d7e483d582183c732a6d4178dfa) was the only required change in order to support `architecture` violations. Adding `visibility` or any other future extension should be as simple as that.

Reply:
Just for the record... all of that reasoning is in line with the work I did on my first PR to this project in https://github.com/rubyatscale/packs/pull/104 where I also removed some hardcoded call to `dependency?` with a more dynamic check, allowing support for both `privacy` and `architecture`.